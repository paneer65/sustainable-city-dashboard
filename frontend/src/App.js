import axios from 'axios'
import React, { useState } from 'react';
import { Router } from 'react-router-dom';
import history from './services/history';
import Routes from './routes';
import { AuthContext } from "./context/auth";
import './App.css';

function App(props) {
  // User state
  const [isAuthenticated, setIsAuthenticated] = useState(false);
  const setAuthenticated = (data) => {
      localStorage.setItem("isAuthenticated", data);
      setIsAuthenticated(data);
    }
  const setJWTToken = (token) => {
    localStorage.setItem("token", token);
  }
  const getJWTToken = () => {
    return localStorage.getItem('token');
  }
  const deleteJWTToken = () => {
    localStorage.removeItem('isAuthenticated');
    localStorage.removeItem('token');
  }

  axios.defaults.baseURL = 'http://127.0.0.1:8000/';
  axios.defaults.headers.post['Content-Type'] = 'application/json';
  axios.defaults.headers.get['Content-Type'] = 'application/json';

  // Add a request interceptor
  axios.interceptors.request.use(function (config) {
    const token = getJWTToken();
    config.headers.common.Authorization = 'Token ' + token;
    return config;
  });

  return (
    <div className="App">
      <AuthContext.Provider value={{
          isAuthenticated, setIsAuthenticated: setAuthenticated, setJWTToken, getJWTToken, deleteJWTToken
        }}>
        <Router history={history}>
          <Routes />
        </Router>
      </AuthContext.Provider>
    </div>
  );
}

export default App;
