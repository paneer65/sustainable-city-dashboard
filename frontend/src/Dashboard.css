.Dashboard{
    display: grid;
    grid-template-columns: 25% 25% 25% 25%;
    grid-template-rows: 200px 700px;
    border: 4px solid black;
  }

.home{
    grid-column:1/5 ;
    grid-row: 1/2;
    border: 2px solid grey;
}

.nav{
    float:right;
}

.bt{
    background-color:white;
    border:2px solid grey;
    color:black;
    border-radius:12px;
    font-size:20px;
    font-family:monospace;
    margin: 25px; 
}


.traffic{
    grid-column:1/2 ;
    grid-row: 2/3;
    border: 2px solid red;
}
.pollution{
    grid-column:2/3 ;
    grid-row: 2/3;
    border: 2px solid blue;
}
.environment{
    grid-column:3/4 ;
    grid-row: 2/3;
    border: solid 2px rebeccapurple;
}
.item6{
    grid-column:4/5 ;
    grid-row: 2/3;
    border: solid 2px darkcyan ;
}