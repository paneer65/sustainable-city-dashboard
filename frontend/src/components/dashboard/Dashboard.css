.Dashboard{
    display: grid;
    grid-template-columns: 25% 25% 25% 25%;
    grid-template-rows: 100px 480px 310px;
    border: 4px solid black;
  }

.home{
    grid-column:1/5 ;
    grid-row: 1/2;
    border: 2px solid grey;
}

.nav{
    margin-right: 50px;
    float:right;
}

.bt{
    background-color:white;
    border:2px solid grey;
    color:black;
    border-radius:12px;
    font-size:20px;
    font-family:monospace;
    margin: 25px; 
}


.traffic{
    grid-column:1/2 ;
    grid-row: 2/3;
}
.pollution{
    grid-column:2/3 ;
    grid-row: 2/3;
}
.environment{
    grid-column:3/4 ;
    grid-row: 2/3;
}
.speed{
    grid-column:4/5 ;
    grid-row: 2/3;
}
.maps{
    grid-column:1/5 ;
    grid-row: 3/4;
}

.bootstrap-card {
    height: 100%;
    border: solid 4px grey ;
    border-radius: 20px;
  }

  .header {
    color: #5e61f6;
    font-family: "Pacifico", cursive;
    font-size: 2.5em;
    letter-spacing: 2px;
    margin-top: -5px;
    text-decoration: none;
    display: inline-block;
}