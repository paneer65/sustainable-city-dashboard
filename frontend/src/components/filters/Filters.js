import React from 'react'
import ListGroup from 'react-bootstrap/ListGroup'
import Button from 'react-bootstrap/Button'
import "./style.css"

const modules_list = ['Traffic', 'Pollution', 'Bikes', 'News', 'Events', 'Bus']

class Filters extends React.Component {
  render() {
    let modules_list_html = modules_list.map((element) =>
      {
        if (this.props.selectedFilter === element) {
          return (
            <ListGroup.Item as="li" key={ element } active >
              <Button variant="dark" onClick={ () => this.props.updateSelectedFilter(element) }>{ element }</Button>
            </ListGroup.Item>
          )
        } else {
          return (
            <ListGroup.Item as="li" key={ element }>
              <Button variant="dark" onClick={ () => this.props.updateSelectedFilter(element) }>{ element }</Button>
            </ListGroup.Item>
          )
        }
      }
    )

    return (
      <div className="Filters">
        <ListGroup as="ul">
          { modules_list_html }
        </ListGroup>
      </div>
    )
  }
}

export default Filters;
