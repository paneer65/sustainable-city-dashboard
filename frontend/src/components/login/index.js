import React, { useState } from "react";
import Form from 'react-bootstrap/Form'
import Button from "react-bootstrap/Button";
import { useAuth } from "../../context/auth";

import "./style.css";

export default function Login(props) {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const { history } = props;
  const { setIsAuthenticated, setJWTToken } = useAuth();

  function validateForm() {
    return username.length > 0 && password.length > 0;
  }

  async function handleSubmit(event) {
    event.preventDefault();

    let data = {
      'username': username,
      'password': password
    }

    fetch('http://localhost:8000/user/login', {
      method: 'POST',
      body: JSON.stringify(data),
      headers: {
        'Content-Type': 'application/json'
      }
    })
    .then(response => {
      if (response.status === 200) {
        return response.json();
      } else if (response.status === 403) {
        setErrorMessage('Username or password is invalid');
        return false;
      } else {
        setErrorMessage('Unknown error')
        return false;
      }
    })
    .then(response => {
      if (response) {
        setIsAuthenticated(true)
        setJWTToken(response['Token'])
        history.push('/dashboard');
      }
    });
  }

  return (
    <div className="Login">
      <Form onSubmit={handleSubmit}>
        <Form.Group controlId="username" size="lg">
          <Form.Label>Username</Form.Label>
          <Form.Control
            autoFocus
            id="username-field"
            type="text"
            value={username}
            onChange={e => setUsername(e.target.value)}
          />
        </Form.Group>
        <Form.Group controlId="password" size="lg">
          <Form.Label>Password</Form.Label>
          <Form.Control
            id="password-field"
            value={password}
            onChange={e => setPassword(e.target.value)}
            type="password"
          />
        </Form.Group>
        {errorMessage}
        <Button block size="lg" disabled={!validateForm()} type="submit">
          Login
        </Button>
      </Form>
    </div>
  );

}
