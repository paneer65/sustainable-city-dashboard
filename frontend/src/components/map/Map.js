import React from "react"
import { compose } from "recompose"
import {
  withScriptjs,
  withGoogleMap,
  GoogleMap,
  Marker,
  InfoWindow
} from "react-google-maps"
import axios from "axios"
import TrafficGoogleMaps from "../traffic/Traffic"
import { Dot } from "recharts"

const defaultLocation = { lat: 53.343786, lng: -6.255828 };
const defaultZoomLevel = 11;
const googleMapURL = "https://maps.googleapis.com/maps/api/js?v=3.exp&libraries=geometry,drawing,places&key=AIzaSyDOjyfAl22KFpq0czq_I0sbRtJHKEkwdIc";
const ozoneWeight = 0.3;
const sulphurDioxideWeight = 0.15;
const nitrogenDioxideWeight = 0.10;
const particleMatter25Weight = 0.20;
const particleMatter10Weight = 0.20;
const carbonMonoxideWeight = 0.5;

class Map extends React.Component {
	constructor(props) {
    super(props)

    this.state = {
		pollutionData: [],
		predictedPollutionData: [],
		bikesData: [],
		busData: [],
		selectedFilter: '',
		pollutionSourcesMerged: false,
		mergedPollutionData: [],
		selectedMarker: '',
		eventsData: []
	}
}

  componentDidMount() {
    if(this.props.selectedFilter === 'Pollution' && this.state.predictedPollutionData.length > 0 && this.state.pollutionData.length > 0 && !this.state.pollutionSourcesMerged){
			this.mergeAndCategorizePollutionData();
		}

    if (this.props.selectedFilter === 'Pollution') {
			this.fetchPollutionMapData();
		} else if (this.props.selectedFilter === 'Bikes') {
			this.fetchBikesMapData();
		} else if (this.props.selectedFilter === 'Events'){
			this.fetchEventsData();
		} else if (this.props.selectedFilter === 'Bus') {
  		this.fetchBusMapData();
  	}
  }

	componentDidUpdate(prevProps) {
		// Need to check state selectedFilter otherwise this will be an infinite loop
		if(this.props.selectedFilter === 'Pollution' && this.state.predictedPollutionData.length > 0 && this.state.pollutionData.length > 0 && !this.state.pollutionSourcesMerged){
			this.mergeAndCategorizePollutionData();
		}
    if (this.props.selectedFilter === 'Pollution' && prevProps.selectedFilter !== 'Pollution') {
			this.fetchPollutionMapData();
		} else if (this.props.selectedFilter === 'Bikes' && prevProps.selectedFilter !== 'Bikes') {
			this.fetchBikesMapData();
		} else if (this.props.selectedFilter === 'Events' && prevProps.selectedFilter !== 'Events'){
			this.fetchEventsData();
		} else if (this.props.selectedFilter === 'Bus' && prevProps.selectedFilter !== 'Bus') {
  		this.fetchBusMapData();
  	}
	}

	updatePollutionAPIData(data) {
		let keys = Object.keys(data);
		let name = data[0]['location_name'];
		let all = []
		let arr = {}
		let count = 0;
		let i =1;

		for(var key in keys)
		{
			if(data[key]['location_name'] === name)
			{
				arr[data[key]['parameter']] = data[key]['value'];
				count++;
			}
			else
			{
				arr['location_name'] = name;
				arr['id'] = i;
				arr['latitude'] = parseFloat(data[key-1]['latitude']);
				arr['longitude'] = parseFloat(data[key-1]['longitude'])
				all.push(arr);
				name = data[key]['location_name'];
				count = 0;
				arr = {}
				i++;
			}
		}
		arr['location_name'] = name;
		arr['id'] = i;
		arr['latitude'] = parseFloat(data[key-1]['latitude']);
		arr['longitude'] = parseFloat(data[key-1]['longitude'])
		all.push(arr)
		return this.airQualityCalculator(all)
	}

	airQualityCalculator(data) {
		let keys = Object.keys(data);
		for ( var key in keys ) {
			var pollutionLevel = 0;
			let dataElement = data[key]
			if (dataElement.hasOwnProperty("o3"))
				pollutionLevel = pollutionLevel + ozoneWeight * dataElement["o3"];
			if (dataElement.hasOwnProperty("so2"))
				pollutionLevel = pollutionLevel + sulphurDioxideWeight * dataElement["so2"];
			if (dataElement.hasOwnProperty("no2"))
				pollutionLevel = pollutionLevel + nitrogenDioxideWeight * dataElement["no2"];
			if (dataElement.hasOwnProperty("pm25"))
				pollutionLevel = pollutionLevel + particleMatter25Weight * dataElement["pm25"];
			if (dataElement.hasOwnProperty("pm10"))
				pollutionLevel = pollutionLevel + particleMatter10Weight * dataElement["pm10"];
			if (dataElement.hasOwnProperty("co"))
				pollutionLevel = pollutionLevel + carbonMonoxideWeight * dataElement["co"];
			data[key]["pollution"] = pollutionLevel
		}
		return data
	}

	mergeAndCategorizePollutionData(){
		let pollutionData = this.state.pollutionData
		let predictedPollutionData = this.state.predictedPollutionData
		let mergedData = []
		for (let key in pollutionData) {
			if (pollutionData.hasOwnProperty(key)) {
				let pollution = pollutionData[key]
				pollution['icon'] = 'red-dot.png'
				pollution['id'] = mergedData.length
				mergedData.push(pollution)
			}
		}
		for (let key in predictedPollutionData) {
			if (predictedPollutionData.hasOwnProperty(key)) {
				let pollution = predictedPollutionData[key]
				pollution['icon'] = 'yellow-dot.png'
				pollution['id'] = mergedData.length
				mergedData.push(pollution)
			}
		}
		for (let key in mergedData) {
			if (mergedData.hasOwnProperty(key)) {
				if(mergedData[key]["pollution"] < 1)
					mergedData[key]["pollutionLevel"] = 'Low'
				else if(mergedData[key]["pollution"] < 2.5)
					mergedData[key]["pollutionLevel"] = 'Average'
				else
					mergedData[key]["pollutionLevel"] = 'High'
			}
		}
		this.setState({ mergedPollutionData: mergedData })
		this.setState({ pollutionSourcesMerged: true})
	}

	fetchPollutionMapData() {
		let updatedData = []
		axios({
			url: '/api/pollution/',
			method: 'GET'
		}).then((response) => {
			if(response.status === 200) {
				updatedData = this.updatePollutionAPIData(response.data)
				this.setState({ pollutionData: updatedData})
			}
		});
		this.fetchPredictedPollutionData()
	}


	eventsTimeKeeper(events) {
		const today = new Date();
		let timeString = today.getHours() + ':' + today.getMinutes() + ':' + today.getSeconds();
		for (let key in events) {
			if (events.hasOwnProperty(key)) {
				let currentEvent = events[key]

				//Events going on now
				if(timeString>currentEvent.starts_at && timeString<currentEvent.ends_at){
					let eventEndingHour = parseInt(currentEvent.ends_at.split(":")[0])
					let currentHour = parseInt(today.getHours())
					if(eventEndingHour-currentHour === 1)
						currentEvent['icon'] = 'green-dot.png'
					else
						currentEvent['icon'] = 'yellow-dot.png'
				}

				//Events going to start in an hour
				else if(timeString<currentEvent.starts_at){
					if(parseInt(currentEvent.starts_at.split(":")[0])-parseInt(today.getHours()) === 1)
						currentEvent['icon'] = 'red-dot.png'
					else
						currentEvent['icon'] = 'ltblue-dot.png'
				}
				else
					currentEvent['icon'] = 'blue-dot.png'
				events[key] = currentEvent
			}
		}
		this.props.updateAllMarkers(events);
		return events
	}


	fetchEventsData() {
		axios({
			url: '/events/',
			method: 'GET'
		}).then((response) => {
			if(response.status === 200) {
				const events = this.eventsTimeKeeper(response.data.events)
				this.setState({ eventsData: events})
			}
		});
	}

	fetchPredictedPollutionData() {
		axios({
		  url: '/api/mlmodel/',
		  method: 'GET'
		}).then((response) => {
		  if(response.status === 200) {
				let predictedPollution = response.data
				this.setState({ predictedPollutionData : predictedPollution })
				//this.forceUpdate();
		  }
		});
	}

  	fetchBikesMapData() {
		axios({
			url: '/api/bikes/',
			method: 'GET'
		}).then((response) => {
			if(response.status === 200) {
				this.setState({ bikesData: response.data })
			}
		});
	}

  fetchBusMapData() {
    axios({
      url: '/api/bus/',
      method: 'GET'
    }).then((response) => {
      if(response.status === 200) {
        this.setState({ busData: response.data })
      }
    });
	}

	handleClick(marker, event) {
		this.setState({ selectedMarker: marker });
    	this.props.updateSelectedMarker(marker);
	}

	googleMapInit() {
		return (
			<GoogleMap defaultZoom = { defaultZoomLevel } defaultCenter = { defaultLocation } />
		)
	}

	generatePollutionMap() {
		return compose(withScriptjs, withGoogleMap)(props => {
		  	return (
				<GoogleMap defaultZoom={ defaultZoomLevel } defaultCenter={ defaultLocation }>
				{
					props.markers.map(marker => {
						const onClick = props.onClick.bind(this, marker)
						return (
							<Marker key = {marker.id}
									onClick={onClick}
									position={{ lat: parseFloat(marker.latitude), lng: parseFloat(marker.longitude) }}
									icon={{url:"http://maps.google.com/mapfiles/ms/icons/" + marker.icon}}>
							{
								props.selectedMarker === marker &&
								<InfoWindow>
									<div>
										<h4>{ marker.pollutionLevel }</h4>
										<span>{ marker.pollution }</span>
									</div>
								</InfoWindow>
							}
							</Marker>
						)
					})
				}
				</GoogleMap>
		  	)
		});
	}

  generateBikesMap() {
		return compose(withScriptjs, withGoogleMap)(props => {
		  	return (
				<GoogleMap defaultZoom={ defaultZoomLevel } defaultCenter={ defaultLocation }>
				{
						props.markers.map(marker => {
							const onClick = props.onClick.bind(this, marker)
							return (
								<Marker	key={marker.location_name}
										onClick={onClick}
										position={{ lat: parseFloat(marker.latitude), lng: parseFloat(marker.longitude) }}>
								{
									props.selectedMarker === marker &&
									<InfoWindow>
										<div>
											<h4>{marker.location_name}</h4>
                      						<span> Number of bikes : {marker.number_of_bikes} </span>
										</div>
									</InfoWindow>
								}
								</Marker>
							)
					})
				}
				</GoogleMap>
		  	)
		});
	}

	generateEventsMap() {
		return compose(withScriptjs, withGoogleMap)(props => {
		  	return (
				<GoogleMap defaultZoom={ defaultZoomLevel } defaultCenter={ defaultLocation }>
				{
						props.markers.map(marker => {
							const onClick = props.onClick.bind(this, marker)
							return (
								<Marker	key={marker.event_name}
										onClick={onClick}
										position={{ lat: parseFloat(marker.latitude), lng: parseFloat(marker.longitude) }}
										icon={{url:"http://maps.google.com/mapfiles/ms/icons/" + marker.icon}}>
								{
									this.props.selectedMarker === marker &&
									<InfoWindow>
										<div>
											<h4>{marker.event_type}</h4>
                      						<span> Name: {marker.event_name} </span><br/>
											<span> Speed Limit: {marker.speed_limit} </span>
										</div>
									</InfoWindow>
								}
								</Marker>
							)
					})
				}
				</GoogleMap>
		  	)
		});
	}

  generateBusMap() {
		return compose(withScriptjs, withGoogleMap)(props => {
		  	return (
				<GoogleMap defaultZoom={ defaultZoomLevel } defaultCenter={ defaultLocation }>
				{
						props.markers.map(marker => {
							const onClick = props.onClick.bind(this, marker)
							return (
								<Marker	key={marker.stop_id + marker.location_name}
										onClick={onClick}
										position={{ lat: parseFloat(marker.latitude), lng: parseFloat(marker.longitude) }}>
								{
									props.selectedMarker === marker
								}
								</Marker>
							)
					})
				}
				</GoogleMap>
		  	)
		});
	}

	generateTrafficMap(){
		return TrafficGoogleMaps
	}

	render() {
		if (this.props.selectedFilter === 'Pollution') {
			const PollutionMap = this.generatePollutionMap();
			//this.mergePollutionData(this.state.pollutionData, this.state.predictedPollutionData)
			return (
				<PollutionMap
				selectedMarker={ this.props.selectedMarker }
				markers={ this.state.mergedPollutionData }
				onClick={ this.handleClick }
				googleMapURL={ googleMapURL }
				loadingElement={<div style= {{ height: `100%` }} />}
				containerElement={<div style= {{ height: `100%` }} />}
				mapElement={<div style = {{ height: `100%` }} />}
    		/>
			)
		}

		else if (this.props.selectedFilter === 'Bikes') {
      		const BikesMap = this.generateBikesMap();
			return (
				<BikesMap
          		selectedMarker={ this.props.selectedMarker }
		      markers={ this.state.bikesData }
		      onClick={ this.handleClick }
		      googleMapURL={ googleMapURL }
		      loadingElement={<div style= {{ height: `100%` }} />}
		      containerElement={<div style= {{ height: `100%` }} />}
		      mapElement={<div style = {{ height: `100%` }} />}
    		/>
			  )
      } else if (this.props.selectedFilter === 'Bus') {
        const BusMap = this.generateBusMap();
        return (
  				<BusMap
            selectedMarker={ this.props.selectedMarker }
  		      markers={ this.state.busData }
  		      onClick={ this.handleClick }
  		      googleMapURL={ googleMapURL }
  		      loadingElement={<div style= {{ height: `100%` }} />}
  		      containerElement={<div style= {{ height: `100%` }} />}
  		      mapElement={<div style = {{ height: `100%` }} />}
      		/>
			  )
    	} else if (this.props.selectedFilter === 'Traffic') {
			const TrafficMap = this.generateTrafficMap();
				  return (
					  <TrafficMap
				selectedMarker={ this.props.selectedMarker }
					markers={ this.state.bikesData }
					onClick={ this.handleClick }
					googleMapURL={ googleMapURL }
					loadingElement={<div style= {{ height: `100%` }} />}
					containerElement={<div style= {{ height: `100%` }} />}
					mapElement={<div style = {{ height: `100%` }} />}
				  />
				  )
		}
		else if (this.props.selectedFilter === 'Events') {
			const EventsMap = this.generateEventsMap();
				return (
					<EventsMap
						selectedMarker={ this.props.selectedMarker }
						markers={ this.state.eventsData }
						onClick={ this.handleClick }
						googleMapURL={ googleMapURL }
						loadingElement={<div style= {{ height: `100%` }} />}
						containerElement={<div style= {{ height: `100%` }} />}
						mapElement={<div style = {{ height: `100%` }} />}
				  	/>
				)
		}
		else {
			const WrappedMap = withScriptjs(withGoogleMap(this.googleMapInit))
			return (
				<WrappedMap
					googleMapURL = { googleMapURL }
					loadingElement={<div style = {{ height: "100%"}} />}
					containerElement={<div style = {{ height: "100%"}} />}
					mapElement={<div style = {{ height: "100%"}} />}
				/>
			)
		}
	}
}

export default Map;
