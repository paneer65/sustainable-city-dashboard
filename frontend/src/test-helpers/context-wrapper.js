import React, { useState } from 'react';

const isAuthenticated = false;
const setIsAuthenticated = false;

export const setAuthenticated = (data) => {
  localStorage.setItem("isAuthenticated", data);
  setIsAuthenticated(data);
}

export const setJWTToken = (token) => {
  localStorage.setItem("token", token);
}

export const getJWTToken = () => {
  return localStorage.getItem('token');
}

export const deleteJWTToken = () => {
  localStorage.removeItem('token');
  localStorage.removeItem('isAuthenticated');
}
