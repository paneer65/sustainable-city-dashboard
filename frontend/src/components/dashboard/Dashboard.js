import React from 'react';
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'

import Map from '../map/Map';
import NewsIndex from '../news';
import LocationLineChart from '../reporting/LocationLineChart';
import NavigationBar from '../navigationbar/NavigationBar';
import Filters from '../filters/Filters';
import FilterActions from '../filter-actions/FilterActions';
import InfoPanelIndex from '../info-panel';

import "./style.css";

class Dashboard extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedFilter: 'Traffic',
      selectedMarker: null,
      allMarkers : []
    };
    this.updateSelectedFilter = this.updateSelectedFilter.bind(this);
    this.updateSelectedMarker = this.updateSelectedMarker.bind(this);
    this.updateAllMarkers = this.updateAllMarkers.bind(this);
  }

  updateSelectedFilter(newFilter) {
    this.setState({
      selectedFilter: newFilter,
      selectedMarker: null
    });
  }

  updateSelectedMarker(newMarker) {
    console.log(newMarker)
    this.setState({ selectedMarker: newMarker });
  }

  updateAllMarkers(updatedMarkers) {
    console.log(updatedMarkers)
    this.setState({ allMarkers: updatedMarkers });
  }

  render() {
    let mainView;
    if (this.state.selectedFilter !== 'News') {
      mainView =  <div style = {{ height: "60vh"}}>
                    <Map
                      selectedFilter={this.state.selectedFilter}
                      selectedMarker={this.state.selectedMarker}
                      updateAllMarkers = {this.updateAllMarkers}
                      updateSelectedMarker={this.updateSelectedMarker}
                    />
                  </div>
    } else {
      mainView = <NewsIndex />
    }

    let infoModule;
    if (this.state.selectedFilter === 'Events') {
      infoModule = <FilterActions selectedFilter={ this.state.selectedFilter } selectedMarker = {this.state.selectedMarker} updateSelectedMarker = { this.updateSelectedMarker } allMarkers = { this.state.allMarkers }/>
    } else {
      infoModule = <InfoPanelIndex selectedFilter={this.state.selectedFilter} selectedMarker={this.state.selectedMarker}/>

    }

    return (
  		<div>
  			<NavigationBar />
        <Container fluid={ true }>
          <Row>
            <Col md={ 2 }>
              <Filters selectedFilter={ this.state.selectedFilter } updateSelectedFilter={ this.updateSelectedFilter }/>
            </Col>
            <Col md={ 7 }>
        			<div className="mainView">
    						{ mainView }
        			</div>
            </Col>
            <Col md={ 3 }>
              { infoModule }
            </Col>
          </Row>
          <Row>
            <div className="reports">
              <LocationLineChart
                selectedMarker={this.state.selectedMarker}
                dataModel={this.state.selectedFilter}
              />
            </div>
          </Row>
        </Container>
  		</div>
    );
  }
}

export default Dashboard
