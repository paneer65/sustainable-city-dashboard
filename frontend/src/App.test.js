import React, { Component } from 'react';
import ReactDOM,{unmountComponentAtNode} from 'react-dom';
import App from './App';
import renderer from 'react-test-renderer';
import Dashboard from './components/dashboard/Dashboard';
import Filters from './components/filters/Filters';
import { MemoryRouter, matchPath }  from 'react-router-dom';
import Map from './components/map/Map';
import { shallow } from 'enzyme';
import * as idx from './components/login/index';
import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
// import sinon from 'sinon';

import { AuthContext } from "./context/auth";
import { setAuthenticated, setJWTToken, getJWTToken, deleteJWTToken } from "./test-helpers/context-wrapper";


Enzyme.configure({
    adapter: new EnzymeAdapter(),
    disableLifecycleMethods: true
});

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  // container *must* be attached to document so events work correctly.
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});


it('dashboard renders correctly', () => {
  const tree = renderer.create(
    <AuthContext.Provider value={{
        isAuthenticated: false, setIsAuthenticated: setAuthenticated, setJWTToken, getJWTToken, deleteJWTToken
      }}>
      <MemoryRouter>
        <Dashboard />
      </MemoryRouter>
    </AuthContext.Provider>
  ).toJSON();
  // expect(tree).toMatchSnapshot();
});


idx.Login = jest.fn();

it('renders login correctly',()=>{

});

it('check state of dashboard filter once rendered', () => {
  let classInstance = shallow(<Dashboard />);
  let initialState = classInstance.state().selectedFilter;
  expect(initialState).toBe('Traffic');
});


// it('check state after button click on dashboard', () => {
//   const mockButtonClick = jest.fn();
//   let classInstance = shallow(<Dashboard />);
//   let initialState = classInstance.state().selectedFilter;
//   console.log(initialState);
//
//   const onButtonClick = sinon.spy();
//   const wrapper = shallow(<Filters onButtonClick={onButtonClick} />);
//   wrapper.find('button').simulate('click');
//   expect(onButtonClick).toHaveProperty('callCount', 1);
//
//   /*
//   let childInstance = shallow(<Filters />);
//   console.log(childInstance.find('.filterButton2'));
//   childInstance.find('.filterButton2').simulate('click');
//
//   let updatedState = classInstance.state().selectedFilter;
//   console.log(updatedState)
//   expect(updatedState).toBe('Pollution');*/
// });
