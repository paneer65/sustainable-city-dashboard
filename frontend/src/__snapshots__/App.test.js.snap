// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard renderes correctly 1`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <a
      className="navbar-brand"
      href="#home"
    >
      Sustainable City Dashboard
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="mr-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rb-event-key="/dashboard"
          href="/dashboard"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          Dashboard
        </a>
      </div>
      <div
        className="navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rb-event-key="/user-management"
          href="/user-management"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          User Management
        </a>
      </div>
      <div
        className="navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rb-event-key="/logout"
          href="/logout"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          Logout
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <div
        className="col-md-2"
      >
        <div
          className="Filters"
        >
          <ul
            className="list-group"
            onKeyDown={[Function]}
          >
            <li
              className="list-group-item active"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <button
                className="filterButton1 btn btn-dark"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Traffic
              </button>
            </li>
            <li
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <button
                className="filterButton2 btn btn-dark"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Pollution
              </button>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="col-md-7"
      >
        <div
          className="maps"
        >
          <div
            style={
              Object {
                "height": "75vh",
              }
            }
          >
            <div
              style={
                Object {
                  "height": "100%",
                }
              }
            />
          </div>
        </div>
      </div>
      <div
        className="col-md-3"
      >
        <div
          className="FilterActions"
        >
          <h2>
            Actions
          </h2>
          <div
            className="TrafficActions"
          >
            Traffic actions
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
